contracts/staking/StakeHolderWIMXV2.sol

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2025
+// SPDX-License-Identifier: Apache 2
+pragma solidity >=0.8.19 <0.8.29;
+
+import {IStakeHolder, StakeHolderBase, StakeHolderBaseV2, StakeHolderNativeV2} from "./StakeHolderNativeV2.sol";
+import {IWIMX} from "./IWIMX.sol";
+
+/**
+ * @title StakeHolderWIMXV2: allows anyone to stake any amount of IMX and to then remove all or part of that stake.
+ * @dev Stake can be added and withdrawn either as native IMX only.
+ * @dev The StakeHolderWIMX contract is designed to be upgradeable.
+ * @dev This contract is the same as StakeHolderWIMX, with the exception that it derives from StakeHolderBaseV2.
+ */
+contract StakeHolderWIMXV2 is StakeHolderNativeV2 {
+    /// @notice The token used for staking.
+    IWIMX internal wIMX;
+
+    /**
+     * @notice Initialises the upgradeable contract, setting up admin accounts.
+     * @param _roleAdmin the address to grant `DEFAULT_ADMIN_ROLE` to
+     * @param _upgradeAdmin the address to grant `UPGRADE_ROLE` to
+     * @param _distributeAdmin the address to grant `DISTRIBUTE_ROLE` to
+     * @param _wIMXToken The address of the WIMX contract.
+     */
+    function initialize(
+        address _roleAdmin,
+        address _upgradeAdmin,
+        address _distributeAdmin,
+        address _wIMXToken
+    ) public initializer {
+        __StakeHolderBase_init(_roleAdmin, _upgradeAdmin, _distributeAdmin);
+        wIMX = IWIMX(_wIMXToken);
+    }
+
+    receive() external payable {
+        // Receive IMX sent by the WIMX contract when wIMX.withdraw() is called.
```

Comment:
consider adding a check that the sender is wIMX to avoid accidentally sent funds being stuck

Reply:
Good point. I have put a check in for this